Tweetinvi.Core/Core/TweetinviSettingsAccessor.cs

```diff
@@ -39,104 +39,135 @@ public interface ITweetinviSettingsAccessor
         RateLimitTrackerMode RateLimitTrackerMode { get; set; }
 
         /// <summary>
-        /// How much additional time to wait than should be strictly necessary for a new batch of Twitter rate limits
-        /// to be available. Required to account for timing discrepancies both within Twitter's servers and between
-        /// Twitter and us. 
+        /// Specify whether you want exceptions coming from Twitter are returning null instead of throwing
         /// </summary>
+        bool OnTwitterExceptionReturnNull { get; set; }
+
         int RateLimitWaitFudgeMs { get; set; }
     }
 
-    public class TweetinviSettingsAccessor : ITweetinviSettingsAccessor
+    public class TweetinviSettingsAccessor : ITweetinviSettingsAccessor, IAsyncContextPreparable
     {
-        private static ITweetinviSettings StaticTweetinviSettings { get; set; }
+        private static ITweetinviSettings StaticTweetinviSettings { get; }
 
-        public TweetinviSettingsAccessor()
+        private static readonly AsyncLocal<ITweetinviSettings> _executionContextTweetinviSettings;
+        private static readonly ThreadLocal<ITweetinviSettings> _threadSettings = new ThreadLocal<ITweetinviSettings>(() => null);
+        private static readonly ThreadLocal<bool> _fromAwait = new ThreadLocal<bool>(() => false);
+
+        static TweetinviSettingsAccessor()
         {
-            var threadSettings = TweetinviCoreModule.TweetinviContainer.Resolve<ITweetinviSettings>();
-            threadSettings.HttpRequestTimeout = 10000;
-            threadSettings.UploadTimeout = 60000;
-            threadSettings.RateLimitWaitFudgeMs = 5000;
+            StaticTweetinviSettings = new TweetinviSettings();
+            _executionContextTweetinviSettings = new AsyncLocal<ITweetinviSettings>();
+        }
 
-            CurrentThreadSettings = threadSettings;
+        public TweetinviSettingsAccessor()
+        {
+            _fromAwait.Value = true;
+            CurrentThreadSettings = new TweetinviSettings();
         }
 
-        [ThreadStatic]
-        private static ITweetinviSettings _currentThreadSettings;
         public ITweetinviSettings CurrentThreadSettings
         {
             get
             {
-                if (_currentThreadSettings == null)
-                {
-                    InitialiseSettings();
-                }
-
-                return _currentThreadSettings;
+                InitializeThreadExecutionContext();
+                return _threadSettings.Value ?? _executionContextTweetinviSettings.Value;
```

Comment:
Shouldn't we use the async local value first and then fall back to thread local?

Reply:
If `_threadSettings` has been set it means that we are in a new Thread created by the user himself.
The other way is not true, `_executionContextTweetinviSettings` comes from `async/await` which might or might not have created a new thread.

Hence the `_threadSettings` are taking priority.

Reply:
Could we not end up on a thread that has an instance of `_threadSettings` set even in the middle of an `async` flow?
e.g. Say we're in a WinForms application, so there's a synchronisation context. The user sets `_threadSettings` on the UI thread (sync context). We then do some `async` work, and during it within Tweetinvi we do `await methodX();` (note there is no `.ConfigureAwait(false);` on the end so we resume on the UI thread due to the synchronisation context), then after that `_threadSettings` is set, even though we could still be within an async call to Tweetinvi.  

This example could be avoided by using `.ConfigureAwait(false)` but I'm very wary of this as I'm sure there will be other strange scenarios where we can end up having `_threadSettings` set where we weren't expecting them to be.

Reply:
I will think about it and try to come up with a test.

If you can come up with a test reproducing this scenario this would be great too.